pkg/health/heartbeat/heartbeat.go

```diff
@@ -310,12 +310,13 @@ func RenewMemberLeasePeriodically(ctx context.Context, stopCh chan struct{}, hco
 
 	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
 	if err != nil {
-		logger.Errorf("failed to create clientset: %v", err)
+		logger.Fatalf("failed to create clientset: %v", err)
```

Comment:
The issue seems to be that while creating a clientSet, api server could not be reached within the timeout time.  Would it make sense to retry with a fixed backoff before completely giving up?
Also it would be a bit cleaner and easier to test if `RenewMemberLeasePeriodically` returns an error. The caller can then choose to exit in case this method returns an error by calling `go log.Fatal(heartbeat.RenewMemberLeasePeriodically(<method arguments>)`

Reply:
It is edge case occurs when `Etcds` are getting updated to newer version. It's about `fail fast` to avoid deadlock.
Retrying with backoff kindof make sense but returning an error doesn’t make sense.
Basically, leader-election is thinking `RenewMemberLeasePeriodically` is running and when leader backup-restore lost election due to some error it will try to close `RenewMemberLeasePeriodically` but it was not running which cause deadlock, so returning an error doesn’t help. check [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/server/backuprestoreserver.go#L211-L222)

Reply:
In the end you need to fail fast. This you can also achieve without putting `logger.Fatalf` inside the `RenewMemberLeasePeriodically` function. This function only returns an error if it has given up. The caller catches the error and then calls `go log.Fatal(heartbeat.RenewMemberLeasePeriodically(<method arguments>)`. In this case `runServer` will return an error and your etcdbr will be forced to fail fast. This is the same pattern that is also followed in `http.ListenAndServe`

Reply:
no, I'm open to change.
I'll try to do it and let you know to review it.

Reply:
Hey @unmarshall ,
I will add this change when I will open a PR on master branch right now I don’t want to make too many changes. Hope that’s fine

Reply:
yes that should be fine.